### PR TITLE
detect/pcre: Use pcre_jit_exec when possible

### PR DESCRIPTION
**Pull request by jlucovsky**
Continuation of #4657

This PR uses `pcre_jit_exec` when available with the `pcre` keyword. It also refactors like code across most of the detect modules into helper functions in `detect-parse.c`

- The refactoring of PCRE related code across dozens of detect modules
- Using `pcre_jit_exec` in the PCRE detect module for recurring patterns

Redmine issue:[749](https://redmine.openinfosecfoundation.org/issues/749)

This update
- Reworks how the JIT stack is used to make it safe within a multi-threaded execution environment.

### CODE REVIEW

File: src/detect-pcre.c [hunk 1]
```diff
@@ -70,15 +71,49 @@
...
+        pcre_jit_stack *jit_stack = (pcre_jit_stack *)
+            DetectThreadCtxGetKeywordThreadCtx(det_ctx, g_jitstack_thread_ctx_id);
+        if (jit_stack) {
+            ret = pcre_jit_exec(regex->regex, regex->study, str, strlen,
+                                start_offset, options, ovector, ovector_size,
+                                jit_stack);
+
+            if (ret == PCRE_ERROR_JIT_STACKLIMIT)
+                SCLogNotice("PCRE JIT stack limit exhausted");
+            if (ret < -2 && ret != -8)
```
Comment (victorjulien) [inline src/detect-pcre.c]: Can you add a comment explaining these magic values?

File: src/detect-pcre.c [hunk 2]
```diff
@@ -70,15 +71,49 @@
...
+    if (det_ctx && g_jitstack_thread_ctx_id != -1) {
+        pcre_jit_stack *jit_stack = (pcre_jit_stack *)
+            DetectThreadCtxGetKeywordThreadCtx(det_ctx, g_jitstack_thread_ctx_id);
+        if (jit_stack) {
+            ret = pcre_jit_exec(regex->regex, regex->study, str, strlen,
+                                start_offset, options, ovector, ovector_size,
+                                jit_stack);
+
+            if (ret == PCRE_ERROR_JIT_STACKLIMIT)
+                SCLogNotice("PCRE JIT stack limit exhausted");
```
Comment (victorjulien) [inline src/detect-pcre.c]: This seems risky, as an attacker might be able to craft input and cause (expensive) output.

File: src/detect-pcre.c [hunk 3]
```diff
@@ -822,6 +836,25 @@ static int DetectPcreParseCapture(const char *regexstr, DetectEngineCtx *de_ctx,
     return -1;
 }
 
+#ifdef PCRE_HAVE_JIT_EXEC
+static void *DetectPcreThreadInit(void *data /*@unused@*/)
+{
+    pcre_jit_stack *jit_stack = pcre_jit_stack_alloc(PCRE_JIT_MIN_STACK, PCRE_JIT_MAX_STACK);
+
+    if (jit_stack == NULL) {
+        SCLogError(SC_ERR_MEM_ALLOC, "unable to allocate PCRE jit stack");
```
Comment (victorjulien) [inline src/detect-pcre.c]: does the pcre API provide a more detailed error output?

File: src/detect-pcre.c [hunk 4]
```diff
@@ -840,6 +873,17 @@ static int DetectPcreSetup (DetectEngineCtx *de_ctx, Signature *s, const char *r
     if (DetectPcreParseCapture(regexstr, de_ctx, pd, capture_names) < 0)
         goto error;
 
+#ifdef PCRE_HAVE_JIT_EXEC
+    if (g_jitstack_thread_ctx_id == -1) {
+        g_jitstack_thread_ctx_id = DetectRegisterThreadCtxFuncs(de_ctx, "jitstack",
+                DetectPcreThreadInit, (void *) &g_jitstack_thread_ctx_id /* unused */,
+                DetectPcreThreadFree, 1);
+        if (g_jitstack_thread_ctx_id == -1) {
+            SCLogNotice("Error while registering jitstack thread funcs");
```
Comment (victorjulien) [inline src/detect-pcre.c]: this should be a `SCLogWarning`

Comment (jlucovsky) [reply src/detect-pcre.c]: Will update.

File: src/detect-pcre.c (hunk 2 above)
Comment (jlucovsky) [reply src/detect-pcre.c]: This is a left over debug message. Will remove.

File: src/detect-pcre.c (hunk 1 above)
Comment (jlucovsky) [reply src/detect-pcre.c]: Left over debug cruft; will remove.

File: src/detect-pcre.c (hunk 3 above)
Comment (jlucovsky) [reply src/detect-pcre.c]: The `pcre2` API has an interface but not `pcre`.

**Comment by jlucovsky** [comment]
Continued in #4693

**Pull request closed without merging**